R/conformal_infer.R

```diff
@@ -0,0 +1,459 @@
+
+  # ------------------------------------------------------------------------------
+  # split new_data for row-wise processing
+
+  new_nest <-
+    dplyr::bind_cols(new_data, new_pred) %>%
+    dplyr::mutate(.row = dplyr::row_number()) %>%
+    dplyr::group_by(.row) %>%
+    tidyr::nest()
+
+  # ------------------------------------------------------------------------------
+  # compute intervals
+
+  if (object$control$method == "grid") {
+    res <- grid_all(new_nest$data, object$wflow, object$training, level, object$control)
+  } else {
+    res <- optimize_all(new_nest$data, object$wflow, object$training, level, object$control)
+  }
+  if(object$control$progress) {
+    cat("\n")
+  }
+  res
+}
+
+# ------------------------------------------------------------------------------
+
+new_conf_infer <- function(wflow, .data, variance, control) {
+  res <-
+    list(
+      wflow = wflow,
+      training = .data,
+      var_model = variance,
+      control = control
+    )
+  class(res) <- "int_conformal_infer"
+  res
+}
+
+check_data <- function(.data, wflow) {
+  hardhat::scream(.data, wflow$blueprint$ptypes$predictors)
+  invisible(NULL)
+}
+
+get_outcome_name <- function(x) {
+  names(x$pre$mold$blueprint$ptypes$outcomes)
+}
+
+get_mode <- function(x) {
+  x %>%
+    hardhat::extract_fit_parsnip() %>%
+    purrr::pluck("spec") %>%
+    purrr::pluck("mode")
+}
+
+check_workflow <- function(x) {
+  if (!workflows::is_trained_workflow(x)) {
+    rlang::abort("'object' should be a fitted workflow object.")
+  }
+  if (get_mode(x) != "regression") {
+    rlang::abort("'object' should be a regression model.")
+  }
+}
+
+# ------------------------------------------------------------------------------
+# Get a rough estimate of the local variance for each new sample being
+# predicted. Aitkin (1987) gives a framework where, assuming normality of
+# residuals, the squared residuals can be modeled using Gamma errors and the
+# log link.
+#
+# This most likely underestimates the local variance so use a multiplier on the
+# variance prediction to make the possible range of the bounds really wide.
+
+var_model <- function(object, train_data) {
+
+  y_name <- get_outcome_name(object)
+
+  train_res <- predict(object, train_data)
+  train_res$resid <- train_data[[y_name]] - train_res$.pred
+  train_res$sq <- train_res$resid^2
+
+  # ------------------------------------------------------------------------------
+  # Get ballpark estimate of range of trial values for new data
+
+  # Use squared residuals as outcome in a gamma model to predict the standard
+  # deviation at a given prediction.
+  var_mod <-
+    try(
+      mgcv::gam(sq ~ s(.pred),
+                data = train_res,
+                family = stats::Gamma(link = "log")),
+      silent = TRUE
+    )
+
+  if (inherits(var_mod, "try-error")) {
+    msg <- c(
+      "!" = "The model to estimate the possible interval length failed with the following message:",
```

Comment:
```suggestion
      "The model to estimate the possible interval length failed with the following message:",
```

My suggestion as-is duplicated the exclamation point.😵‍💫